core/src/main/java/com/datastax/oss/driver/api/core/config/ProgrammaticConfigBuilder.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.api.core.config;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * A builder that allows the creation of a config loader where options are overridden
+ * programmatically.
+ *
+ * @see DriverConfigLoader#programmaticBuilder()
+ */
+public interface ProgrammaticConfigBuilder extends OngoingConfigOptions<ProgrammaticConfigBuilder> {
+
+  /**
+   * Starts the definition of a new profile.
+   *
+   * <p>All options set after this call, and before the next call to this method or {@link
+   * #endProfile()}, will apply to the given profile.
+   */
+  @NonNull
+  ProgrammaticConfigBuilder startProfile(@NonNull String profileName);
+
+  /**
+   * Ends the definition of a profile. All options set after this call, and before the next call to
+   * {@link #startProfile(String)}, will apply to the default profile.
+   */
+  @NonNull
+  ProgrammaticConfigBuilder endProfile();
```

Comment:
Why do we need `endProfile` name? When the user creates a new profile it can use `startProfile(p1)`. Then when the user doesn't want to create another profile maybe it is enough to call `build`?
Otherwise, if he wants to define another profile he can call `startProfile(p2)` - in that way we are simplifying the API because there is no need to call `endProfile`

Reply:
From the current implementation I see that endProfile() does:
`currentProfileName = DriverExecutionProfile.DEFAULT_NAME;`
So shouldn't it be named: `backToDefaultProfile` or smth similar?

Reply:
> So shouldn't it be named: backToDefaultProfile or smth similar?

That's pretty much the same isn't it?
If you look at the structure of the file, you have the default profile at the root level, and then the `profiles` section. So it should be pretty clear that ending a profile goes back to the default profile. And if not, that's documented in the javadoc.

Reply:
To be honest, that behavior of automatically getting back to default profile was not clear to me. 
I wonder if our users will not have a similar experience. 
I was using similar builders from Spring framework for configuration for example security and it was more explicit so maybe that is why it seems not intuitive to me.